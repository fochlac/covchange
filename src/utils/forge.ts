export const forge = {
	repositoryId,
	branchId,
	pullRequestId,
}

function repositoryId({ type, project, repo }: Core.Repository): Core.Id {
	return `${type}_${project}_${repo}`
}

function branchId({ name, repository }: Core.Branch | Core.BranchRest | { name: string; repository: Core.Repository }): Core.Id {
	return `${repositoryId(repository)}_${name}`
}

function pullRequestId({ name, repository }: { name: string; repository: Core.Repository }): Core.Id {
	return `${repositoryId(repository)}_${name}`
}
