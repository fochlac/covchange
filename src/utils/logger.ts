import { createWriteStream, existsSync, mkdirSync } from 'fs'

/**
 *   Log Levels:
 *   0: no error, very important info
 *   1: error: critical error in app
 *   2: error: internal error in app
 *   3: error: minor error
 *
 *   4: info: important info (bad request/login)
 *   5: info: minor info
 *
 *   6: trace: tracing informations
 *   7: unimportant tracing information
 *
 *  10: full trace: (requests objects, etc)
 *
 **/
let currentDate = new Date().toISOString().split('T')[0]
if (!existsSync(global.logFolder)) {
	mkdirSync(global.logFolder)
}

let logStream = createWriteStream(global.logFolder + '/covwatcher_' + currentDate + '.log', {
	flags: 'a',
	encoding: 'utf8',
	autoClose: true,
})

logStream.on('error', err => {
	console.log(err)
})

export default (level: number, ...message: any[]): void => {
	const now = new Date().toISOString()
	const messages = message.map(parseMessage).join(' ')
	const logMessage = `${now} - ${level} - ${messages}`

	if (level <= global.logLevel) {
		console.log(logMessage)
	}

	if (currentDate !== new Date().toISOString().split('T')[0]) {
		currentDate = new Date().toISOString().split('T')[0]
		logStream.end()
		logStream = createWriteStream(global.logFolder + '/covwatcher_' + currentDate + '.log', {
			flags: 'a',
			encoding: 'utf8',
			autoClose: true,
		})
	}

	logStream.write(logMessage + '\n')
}

function parseMessage(item) {
	let output
	if (['string', 'number', 'boolean'].includes(typeof item)) {
		output = item
	} else {
		try {
			if (Object.prototype.toString.call(item) === '[object Error]') {
				output = '\n' + item.stack.toString()
			} else {
				output = '\n' + JSON.stringify(item)
			}
		} catch (err) {
			console.log(err + '\n' + item)
			output = ''
		}
	}
	return output
}
