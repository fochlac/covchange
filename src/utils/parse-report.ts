import { parse } from 'fast-xml-parser'

const config = {
	ignoreAttributes: false,
	attrNodeName: 'attr',
	attributeNamePrefix: '',
	parseAttributeValue: true,
}

/**
 * Function to parse a raw clover coverage report in xml format
 *
 * @param {string} rawXML coverage report as utf8 string
 * @returns {Core.Report} parsed coverage report
 */
export function createReportFromXML(rawXML: string): Core.Report {
	const rawProject = parse(rawXML, config).coverage.project

	const folders = rawProject.metrics.package.reduce(reduceFolder, {})
	const files = Object.values(folders).reduce(reduceFiles, {})

	return {
		timestamp: rawProject.attr.timestamp,
		metrics: parseMetrics(rawProject.metrics.attr),
		folders,
		files,
	}
}

function reduceFolder(folders, folder) {
	folders[folder.attr.name] = {
		name: folder.attr.name,
		metrics: parseMetrics(folder.metrics.attr),
		files: parseFile(folder.file),
	}
	return folders
}

function reduceFiles(files, folder) {
	folder.files.forEach(file => {
		files[`${folder.name.split('.').join('/')}/${file.name}`] = file
	})
	return files
}

function parseFile(file) {
	const rawFiles = file && file.length ? file : file ? [file] : []

	return rawFiles.map(file => ({
		name: file.attr.name,
		metrics: parseMetrics(file.metrics.attr),
	}))
}

function parseMetrics({ statements, coveredstatements, conditionals, coveredconditionals }) {
	return {
		statements,
		conditionals,
		coveredStatements: coveredstatements,
		coveredConditionals: coveredconditionals,
		statementCov: (statements && Math.round((coveredstatements / statements) * 10000)) / 100 || 0,
		conditionalCov: (conditionals && Math.round((coveredconditionals / conditionals) * 10000)) / 100 || 0,
	}
}
