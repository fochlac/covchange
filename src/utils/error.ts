import logger from './logger'

const log = (level, ...message) => logger(level, 'error -', ...message)

const error = (location: string): Core.ErrorConstructor => ({
	internalError: (level: number, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)
	},
	routerError: (level: number, res: Express.Response, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)

		res.status(500).send({ success: false })
	},
	badRequest: (level: number, res: Express.Response, field: string, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)

		res.status(403).send({ success: false, field })
	}
})

export default error
