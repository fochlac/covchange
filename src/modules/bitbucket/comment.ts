import * as request from 'request-promise-native'

import { commentDb } from '../db/comments'
import { diffReports } from '../diff/diff'
import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const Authorization = 'Bearer OTQyODM5NTkyMzQ2OqjtV+lhUZPQdkGwO7kqarTLMxrB'
const baseUrl = ({ repository: { slug, name }, pullRequestId }: Core.PullRequest) =>
	`${global.bitbucket}/rest/api/1.0/${slug}/repos/${name}/pull-requests/${pullRequestId}/comments/`

const symbolsMap = {
	critical: ':u7981:',
	low: ':u6709:',
	medium: ':symbols:',
	high: ':u6307:',
}

const getSymbolFromCoverage = (metrics: Core.Metrics) => {
	const { statementCov, statements } = metrics
	const stringcode = statementCov > 85 ? 'high' : statementCov > 70 || statements <= 5 ? 'medium' : statementCov > 50 ? 'low' : 'critical'
	return symbolsMap[stringcode]
}

export function createComment(base: Core.Branch, pr: Core.PullRequest) {
	const diffLatest = diffReports(base.reports[0], pr.reports[0])

	return sendComment(pr, createCommentObject(diffLatest))
}

async function sendComment(pr, comment) {
	const existingComment = await commentDb.get(pr.repository, pr.pullRequestId)
	const uri = baseUrl(pr) + (existingComment ? existingComment.commentId : '')
	const method = existingComment ? 'PUT' : 'POST'

	return request({
		uri,
		method,
		json: true,
		headers: {
			Authorization,
		},
		body: comment,
	}).catch(internalError(2, 'Error posting comment.'))
}

function createCommentObject(diff: Core.DiffReport) {
	const totalDiff = diff.total.diff.statementCov
	const changedFilesHeader = (Object.keys(diff.changed).length && ['| File | Change | Coverage |', '|---|---|---|']) || []
	const deletedFilesHeader = (diff.deleted.length && ['##### Deleted files', '```diff']) || []
	const deletedFilesFooter = (diff.deleted.length && ['```']) || []

	const lines = [
		'### Coverage Statistics',
		`#### ${totalDiff <= 0 ? ':red_circle:' : ':large_blue_circle:'} This pull request will ` +
			`${totalDiff > 0 ? 'increase' : 'decrease'} total coverage by ${Math.abs(totalDiff)}% to ${diff.total.changed.statementCov}%.`,
		'',
		...changedFilesHeader,
		...Object.keys(diff.changed)
			.sort((a, b) => diff.changed[a].diff.statementCov - diff.changed[b].diff.statementCov)
			.slice(0, 10)
			.map(name => changedFileInfo(name, diff.changed[name])),
		...Object.keys(diff.new)
			.sort((a, b) => diff.new[a].statementCov - diff.new[b].statementCov)
			.slice(0, 10)
			.map(name => newFileInfo(name, diff.new[name])),
		...deletedFilesHeader,
		...diff.deleted.map(deletedFileInfo),
		...deletedFilesFooter,
	]

	return { text: lines.join('\n') }
}

function newFileInfo(name: string, metrics: Core.Metrics) {
	return `| ${name} | :sparkles: *new* | ${getSymbolFromCoverage(metrics)} ${metrics.statementCov}% |`
}

function deletedFileInfo(name: string) {
	return `- ${name}`
}

function changedFileInfo(name: string, diff: Core.Diff) {
	const covSymbol = diff.diff.statementCov < 0 ? ':red_circle:' : ':large_blue_circle:'
	return `| ${name} | ${covSymbol} ${diff.diff.statementCov}% | ${getSymbolFromCoverage(diff.changed)} ${diff.changed.statementCov}% |`
}
