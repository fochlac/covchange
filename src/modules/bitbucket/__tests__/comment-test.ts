import { createReportFromXML } from '../../../utils/parse-report'
import { readFileSync } from 'fs'
import request from 'request-promise-native'
import { submitResults } from '../comment'

jest.mock('request-promise-native', () =>
	require
		.requireActual('sinon')
		.stub()
		.resolves(),
)

jest.mock('../../../utils/logger.ts', () => () => null)

describe('diff-module', () => {
	const report = createReportFromXML(readFileSync('./tests/sample_report.xml', { encoding: 'utf8' }))
	const report2 = createReportFromXML(readFileSync('./tests/sample_report_2.xml', { encoding: 'utf8' }))
	const repository = {
		repo: 'name',
		project: 'slug',
		type: 'users',
	}
	const base = {
		name: 'base',
		reports: [report],
		repository,
		id: '1',
		lastModified: 0,
	}
	const pr = {
		reports: [report2],
		repository,
		name: '12345',
		id: '1',
		lastModified: 0,
		base: {
			name: 'base',
			repository,
		},
	}

	it('should find the diff between the reports', () => {
		return submitResults(base, pr).then(() => {
			expect(request.callCount).toEqual(1)
			expect(request.firstCall.args).toMatchSnapshot()
		})
	})
})
