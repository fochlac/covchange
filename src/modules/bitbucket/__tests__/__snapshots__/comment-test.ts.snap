// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff-module should find the diff between the reports 1`] = `
Array [
  Object {
    "body": Object {
      "text": "### Coverage Statistics
#### :large_blue_circle: This pull request will increase total coverage by 2.4% to 87.2%.

| File | Change | Coverage |
|---|---|---|
| orders/app/components/orders-subviews.js | :red_circle: -13% | :u6709: 68% |
| orders/app/index.js | :large_blue_circle: 7.7% | :u7981: 7.7% |
| orders/app/components/orders-filter-bar.js | :large_blue_circle: 12.8% | :u6307: 96.4% |
| orders/app/components/order-tours-tab/order-tours-table-renderer.js | :large_blue_circle: 44.4% | :u7981: 44.4% |
| orders/app/components/order-tours-tab/order-tours-table-core.js | :sparkles: *new* | :u6709: 54.5% |
##### Deleted files
\`\`\`diff
- orders/module.js
\`\`\`",
    },
    "headers": Object {
      "Authorization": undefined,
    },
    "json": true,
    "method": "post",
    "uri": "undefined/rest/api/1.0/users/name/repos/slug/pull-requests/12345/comments/",
  },
]
`;
